st2actions/tests/unit/test_worker.py

```diff
@@ -71,19 +84,41 @@ def test_non_utf8_action_result_string(self):
             action_worker._run_action(liveaction_db)
         except InvalidStringData:
             liveaction_db = LiveAction.get_by_id(liveaction_db.id)
-            self.assertEqual(liveaction_db.status, "failed")
+            self.assertEqual(liveaction_db.status, action_constants.LIVEACTION_STATUS_FAILED)
             self.assertTrue('error' in liveaction_db.result)
             self.assertTrue('traceback' in liveaction_db.result)
             execution_db = ActionExecution.get_by_id(execution_db.id)
-            self.assertEqual(liveaction_db.status, "failed")
+            self.assertEqual(liveaction_db.status, action_constants.LIVEACTION_STATUS_FAILED)
 
-    def _get_liveaction_model(self, action_db, params):
-        status = action_constants.LIVEACTION_STATUS_REQUESTED
-        start_timestamp = date_utils.get_datetime_utc_now()
-        action_ref = ResourceReference(name=action_db.name, pack=action_db.pack).ref
-        parameters = params
-        context = {'user': cfg.CONF.system_user.user}
-        liveaction_db = LiveActionDB(status=status, start_timestamp=start_timestamp,
-                                     action=action_ref, parameters=parameters,
-                                     context=context)
-        return liveaction_db
+    def test_worker_shutdown(self):
+        action_worker = actions_worker.get_worker()
+
+        # Create a temporary file that is deleted when the file is closed and then set up an
+        # action to wait for this file to be deleted. This allows this test to run the action
+        # over a separate thread, run the shutdown sequence on the main thread, and then let
+        # the local runner to exit gracefully and allow _run_action to finish execution.
+        with tempfile.NamedTemporaryFile() as fp:
+            params = {'cmd': 'while [ -e \'%s\' ]; do sleep 1; done' % fp.name}
+            liveaction_db = self._get_liveaction_model(WorkerTestCase.local_action_db, params)
+            liveaction_db = LiveAction.add_or_update(liveaction_db)
+            executions.create_execution_object(liveaction_db)
+            runner_thread = eventlet.spawn(action_worker._run_action, liveaction_db)
+
+            # Wait for the worker to add the liveaction to _running_liveactions.
+            while len(action_worker._running_liveactions) <= 0:
+                eventlet.sleep(1)
```

Comment:
Do something like 0.1. Otherwise, this test might take foreva. Unfortunately, I don't see a way to test this without sleep.